server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+        return new SnapshotShardSizeInfo(knownSnapshotShardSizes);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            final Set<SnapshotShard> onGoingSnapshotRecoveries = new HashSet<>();
+
+            int unknownShards = 0;
+            for (ShardRouting shardRouting : event.state().routingTable().shardsWithState(ShardRoutingState.UNASSIGNED)) {
+                if (shardRouting.primary() && shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    final RecoverySource.SnapshotRecoverySource snapshotRecoverySource =
+                        (RecoverySource.SnapshotRecoverySource) shardRouting.recoverySource();
+                    final SnapshotShard snapshotShard = new SnapshotShard(snapshotRecoverySource.snapshot(),
+                        snapshotRecoverySource.index(), shardRouting.shardId());
+                    onGoingSnapshotRecoveries.add(snapshotShard);
+
+                    // check if already populated entry
+                    if (knownSnapshotShardSizes.containsKey(snapshotShard) == false) {
+                        // check if already fetching snapshot info in progress
+                        if (unknownSnapshotShards.add(snapshotShard)) {
+                            queue.add(snapshotShard);
+                            unknownShards += 1;
+                        }
+                    }
+                }
+            }
+
+            // Clean up keys from knownSnapshotShardSizes that are no longer needed for recoveries
+            synchronized (this) {
+                isMaster = true;
+                ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = null;
+                for (ObjectCursor<SnapshotShard> shard : knownSnapshotShardSizes.keys()) {
+                    if (onGoingSnapshotRecoveries.contains(shard.value) == false) {
+                        if (newSnapshotShardSizes == null) {
+                            newSnapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                        }
+                        newSnapshotShardSizes.remove(shard.value);
+                    }
+                }
+                if (newSnapshotShardSizes != null) {
+                    knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                }
+            }
+
+            final int nbFetchers = Math.min(unknownShards, maxConcurrentFetches);
+            for (int i = 0; i < nbFetchers; i++) {
+                final int activeFetches = concurrentFetches.incrementAndGet();
+                if (activeFetches < maxConcurrentFetches + 1) {
+                    fetchNextSnapshotShard();
+                } else {
+                    final int value = concurrentFetches.decrementAndGet();
+                    assert value >= 0 : "Unexpected value: " + value;
+                }
+            }
+        } else {
```

Comment:
NIT: `} else if (event.previousState().nodes().isLocalNodeElectedMaster()) {` maybe so we don't execute this over and over? 
I also wonder if we even need to do this? Maybe we should instead just clean out unnecessary items in the map on `event.routingTableChanged()` if we're not master to not have to repopulate the map over and over in an unstable master situation where it might jump in and out of the cluster a few times (which would be really painful if we're dealing with a large number of shards)?

Reply:
+1 on the previous state isLocalNodeElectedMaster() condition (ada6dd50fd6c550822361263efe8e943bae6e13f).

I added a `//TODO` for the other point you raise. I'm not sure we should do it but if we think we should I can do it in a follow up PR.